### ISSUE TITLE
apply str() to the password field of URL

### ISSUE THREAD

**Issue by leandropls** [sqlalchemy/sqlalchemy#6958]
### Describe the bug

On the [documents](https://docs.sqlalchemy.org/en/14/core/engines.html?highlight=url#sqlalchemy.engine.URL), it's said that the password attribute may be any object with a `__str__` method. 

While this interface works with psycopg2, it doesn't with asyncpg. Checking `asyncpg.connection.connect`, it requires either a string or a callable that returns a string.

Passing a password object that implements both `__str__` and `__call__` makes it work.

### To Reproduce

```python
import asyncio
from dataclasses import dataclass

from sqlalchemy.ext.asyncio import create_async_engine
from sqlalchemy.engine import URL

@dataclass
class DynamicPassword
    password: str

    def __str__(self) -> str:
        return self.password

async def async_main():
    engine = create_async_engine(
        URL.create(
            drivername="postgresql+asyncpg",
            username="scott",
            password=DynamicPassword(password="tiger"),
            host="localhost",
            port=5432,
            database="test",
        ), echo=True,
    )

    await engine.execute("select 1")
    await engine.dispose()

asyncio.run(async_main())
```

### Error

```
Traceback (most recent call last):
  ...
  File ".../env/lib/python3.9/site-packages/sqlalchemy/future/engine.py", line 419, in connect
    return super(Engine, self).connect()
  File ".../env/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 3165, in connect
    return self._connection_cls(self, close_with_result=close_with_result)
  File ".../env/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 96, in __init__
    else engine.raw_connection()
  File ".../env/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 3244, in raw_connection
    return self._wrap_pool_connect(self.pool.connect, _connection)
  File ".../env/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 3211, in _wrap_pool_connect
    return fn()
  File ".../env/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 307, in connect
    return _ConnectionFairy._checkout(self)
  File ".../env/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 767, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File ".../env/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 425, in checkout
    rec = pool._do_get()
  File ".../env/lib/python3.9/site-packages/sqlalchemy/pool/impl.py", line 256, in _do_get
    return self._create_connection()
  File ".../env/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 253, in _create_connection
    return _ConnectionRecord(self)
  File ".../env/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 368, in __init__
    self.__connect()
  File ".../env/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 611, in __connect
    pool.logger.debug("Error on connect(): %s", e)
  File ".../env/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
    compat.raise_(
  File ".../env/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
  File ".../env/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 605, in __connect
    connection = pool._invoke_creator(self)
  File ".../env/lib/python3.9/site-packages/sqlalchemy/engine/create.py", line 578, in connect
    return dialect.connect(*cargs, **cparams)
  File ".../env/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 584, in connect
    return self.dbapi.connect(*cargs, **cparams)
  File ".../env/lib/python3.9/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 747, in connect
    await_only(self.asyncpg.connect(*arg, **kw)),
  File ".../env/lib/python3.9/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 67, in await_only
    return current.driver.switch(awaitable)
  File ".../env/lib/python3.9/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 120, in greenlet_spawn
    value = await result
  File ".../env/lib/python3.9/site-packages/asyncpg/connection.py", line 2045, in connect
    return await connect_utils._connect(
  File ".../env/lib/python3.9/site-packages/asyncpg/connect_utils.py", line 790, in _connect
    raise last_error
  File ".../env/lib/python3.9/site-packages/asyncpg/connect_utils.py", line 776, in _connect
    return await _connect_addr(
  File ".../env/lib/python3.9/site-packages/asyncpg/connect_utils.py", line 676, in _connect_addr
    return await __connect_addr(params, timeout, True, *args)
  File ".../env/lib/python3.9/site-packages/asyncpg/connect_utils.py", line 726, in __connect_addr
    await compat.wait_for(connected, timeout=timeout)
  File ".../env/lib/python3.9/site-packages/asyncpg/compat.py", line 66, in wait_for
    return await asyncio.wait_for(fut, timeout)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/tasks.py", line 494, in wait_for
    raise exceptions.TimeoutError() from exc
asyncio.exceptions.TimeoutError
```

### Versions

- OS: reproduced on Debian Bullseye and MacOS 11
- Python: 3.9
- SQLAlchemy: 1.4
- Database: PostgreSQL
- DBAPI: asyncpg

**Comment by CaselIT** [comment]
thanks for reporting

**Comment by CaselIT** [comment]
I guess this may also happen in other dialect.

> password attribute may be any object with a str method.

@zzzeek What was the use case for this?

**Comment by zzzeek** [comment]
someone had some kind of secure password object that wouldn't repr() , or something like that

**Comment by zzzeek** [comment]
im not sure why we dont have tests for this.  there should be a general backend test that runs get_connect_args(), in test/engine/test_parseconnect.py, given a password object

**Comment by CaselIT** [comment]
do we want to keep the use case? or should be just call `str(url.password)` before injecting the args to the dialect?

> im not sure why we dont have tests for this. there should be a general backend test that runs get_connect_args(), in test/engine/test_parseconnect.py, given a password object

maybe it's not `backend=true`

**Comment by leandropls** [comment]
In my specific case, I was setting up a dynamic IAM password at AWS.

This feature made it extremely easy (code a bit simplified):

```
@dataclass
class IAMPassword:
    db_username: str
    db_host: str
    db_port: int
    db_region: str

    def __str__(self) -> str:
        client = boto3.client("rds", self.db_region)
        return client.generate_db_auth_token(
            DBHostname=self.db_host,
            Port=str(self.db_port),
            DBUsername=self.db_username,
            Region=self.db_region,
        )

    def __call__(self) -> str:
        return str(self)

```

**Comment by CaselIT** [comment]
Ok, thanks for the use case example. 

Also careful that iirc boto3 is blocking, so that snipped would block the asyncio loop if used in an async context

**Comment by leandropls** [comment]
@CaselIT thanks for the warning. The token expires every 900s, so I've added some cache that only expires every 840s. My thinking is that under this condition the impact will be negligible, but I'll be watching how it behaves on the server. If it causes any trouble, I'll put the renewal on the event loop.

**Comment by CaselIT** [comment]
iirc there are boto3 asyncio wrappers, but I've never used any (not boto3) directly, so I cannot suggest any.

**Comment by leandropls** [comment]
Yeah... I didn't try any of them because putting it on the event loop seemed like an over-engineered solution for something that blocks only for such a small percentage of the time.

**Comment by francois-gay** [comment]
I confirm that the bug is also present with pg8000.
I've opened an issue in their GitHub (https://github.com/tlocke/pg8000/issues/81), and think this modification should be done in SQLAlchemy.

My use case is exactly the same as @leandropls 's one.

**Comment by CaselIT** [comment]
> 
> 
> im not sure why we dont have tests for this. there should be a general backend test that runs get_connect_args(), in test/engine/test_parseconnect.py, given a password object

we do test this, but each dialect should handle this manually in `connect` since there is no other call that receives the connection arguments at each connection.

I'm inclined to suggest updating the documentation to suggest subclassing string when providing a custom class.

@leandropls @francois-gay  would subclassing str in your use cases solve this issue?

**Comment by leandropls** [comment]
I don't see how this would work, since str is immutable.

**Comment by CaselIT** [comment]
A subclass is not:

```py
class MyStr(str):
    def __init__(self):
        self.c = 0

    def __str__(self):
        self.c += 1
        return f"v={self.c}"

    __repr__ = __str__


s = MyStr()
print(s, s, s, s)
```
outputs `v=1 v=2 v=3 v=4`

**Comment by zzzeek** [comment]
I'm traveling and don't have time to look but it should absolutely be possible for the defaultdialect to be adjusted in this way as well as to provide a suite test to ensure all dialects do the right thing.  I would not be in favor of subclassing string as that's an unusual and awkward pattern.

**Comment by CaselIT** [comment]
I think we would need to add a "pre_connect" call that takes the current connect arguments and returns them.

The dialect do have a `create_connect_args` but that's called once per engine, so that cannot be used here. Also "create_connect_args" may have renamed the password argument into something else, so it complicates doing this in a general way.

>  I would not be in favor of subclassing string as that's an unusual and awkward pattern.

Providing a base class, something like `MutablePassword` may be an option

**Comment by leandropls** [comment]
> A subclass is not:

I'd never guess that printing and formatting a `str` would call `__str__`. I think I've never seen this before. Thank you!

I wouldn't do it this way on the grounds that this is too obscure (sorta like what @zzzeek said), but if you choose to go this path, I'd suggest specifically providing this recipe for cloud users.

I don't see how providing a special base class would be better than just supporting anything that provides a `__str__` method. The way currently documented seems pretty good to me.

**Comment by CaselIT** [comment]
> I don't see how providing a special base class would be better than just supporting anything that provides a __str__ method. The way currently documented seems pretty good to me.

this base class would be a `str` subclass, so that user code would not need to know that str can be subclasses/modified/etc

> In fact, the whole url, IMV, could accept this interface.

This issue is that the url is not passed to the dialect. It's parsed into args and kwargs that are custom for each dialect, and this parsing is only done once.

**Comment by leandropls** [comment]
> this base class would be a `str` subclass, so that user code would not need to know that str can be subclasses/modified/etc

> This issue is that the url is not passed to the dialect. It's parsed into args and kwargs that are custom for each dialect, and this parsing is only done once.

Ahh, I see. Because the URL object is created out of the `connect` function defined on engine/create.py#L572. So to support this, the whole URL parsing would have to be moved inside of the connect function.

If I understand it correctly, if there's a place where this dynamic behavior could be implemented, it'd be inside of `URL.__new__` class, by replacing non-str instances with a str subclass like the one you proposed. Though that would violate  PEP 20's "Explicit is better than implicit".

I see your point now. Despite it's awkwardness, it seems like a pretty good solution to me now, even allowing for even more dynamic fields on the URL object.

**Comment by francois-gay** [comment]
Creating a subclass of str could be a bypass, but it doesn't match the documentation:

> URL.password: password, which is normally a string but may also be any object that has a __str__() method.

Did you see the answers in [my pg8000 issue](https://github.com/tlocke/pg8000/issues/81) ? What do you think about it ?

@leandropls :
> even allowing for even more dynamic fields on the URL object

👍

**Comment by leandropls** [comment]
> Did you see the answers in [my pg8000 issue](https://github.com/tlocke/pg8000/issues/81) ? What do you think about it ?

It's more complicated than they think. Because the arguments are sent to the driver at [engine/default.py#L582](https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/engine/default.py#L582), which runs everytime SQLAlchemy's engine connects to the database. But the arguments aren't really created there, but at [engine/create.py#L564](https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/engine/create.py#L564), at a much earlier stage -- when the engine gets created.

The arguments could be hacked later to str() them before passing them to the driver, but it'd likely become a source of bugs and maintenance problems in the future. That's the beauty of the str subclass... the argument could be passed unchanged from application code, through SQLAlchemy's code, into the driver's code unchanged.

I agree that either the behavior should fixed (most likely on `URL.__new__`) or the documentation changed.

**Comment by CaselIT** [comment]
> 
> 
> Creating a subclass of str could be a bypass, but it doesn't match the documentation:
> 
> > URL.password: password, which is normally a string but may also be any object that has a **str**() method.
> 
> Did you see the answers in [my pg8000 issue](https://github.com/tlocke/pg8000/issues/81) ? What do you think about it ?
> 
> @leandropls :
> 
> > even allowing for even more dynamic fields on the URL object
> 
> 👍

Thanks for pointing that out. Looking at how they use the password, a simple string subclass would not work, since using my example above `s.encode() == b''`.

I guess a new dialect method is needed. Maybe we could skip if it we detect that's not needed, since I would guess the vast majority of usages have static connect arguments, and it would add a superfluous function call in an hot path.

**Comment by francois-gay** [comment]
What do you think about changing the [connect method](https://github.com/sqlalchemy/sqlalchemy/blob/06f031e55f40403996e3580bdc5a22661b6303f4/lib/sqlalchemy/engine/default.py#L582) like that ?

```python
    def connect(self, *cargs, **cparams):
        # inherits the docstring from interfaces.Dialect.connect
        if 'password' in cparams and cparams['password'] is not None and not isinstance(cparams['password'], (str, bytes)):
            cparams['password'] = str(cparams['password'])
        return self.dbapi.connect(*cargs, **cparams)
```

**Comment by CaselIT** [comment]
That's an option, but as mentioned above this would be a change that impact each dialect that overrides the connect method and/or renames the `password` key in `cparams` or otherwise manipulates it in `create_connect_args`.

@zzzeek do you have suggestions mike?

**Comment by francois-gay** [comment]
I agree, but it could treat the majority of dialects, isn't it?

Else : define an abstract DynamicConnectParameter class in SQLAlchemy ant treat it at the same place:

```python
    def connect(self, *cargs, **cparams):
        # inherits the docstring from interfaces.Dialect.connect
        for param in cparams:
            if isinstance(cparams['param'], DynamicConnectParameter):
                cparams['param'] = str(cparams['param'])
        return self.dbapi.connect(*cargs, **cparams)
```

and use it like that : 

```python
class ApplicationDefaultTokenPassword(DynamicConnectParameter):
    # set un instance of this class as database password
    def __init__(self) -> None:
        self.token = None
        self.expiry = None

    def __str__(self) -> str:
        if self.token is None or self.expiry < datetime.utcnow():
            scopes = ['https://www.googleapis.com/auth/cloud-platform']
            credentials, _  = default_credentials(scopes=scopes)
            request = requests.Request()
            credentials.refresh(request)
            self.token = credentials.token
            self.expiry = credentials.expiry
        
        return self.token
```

**Comment by CaselIT** [comment]
sure something like this may work, but should also support the use case where an user can provide custom "creators" to the pool.

I'm not sure if we are overthinking this

**Comment by francois-gay** [comment]
I thought about it: if a user uses the "creator" option, he manages the parameters himself, and doesn't need this use case.
Isn't it ?

**Comment by zzzeek** [comment]
hey there -

"creator" is not part of this, because it has nothing to do with parsing a URL.

The process by which the URL becomes broken into dialect-specific arguments is mostly standardized, and the patch below solves the issue for all dialects that use translate_connect_args(), which should be all of them.   If we can add a suite test for all of our dialects which will also cover third party, that will resolve the issue:

```py
diff --git a/lib/sqlalchemy/engine/url.py b/lib/sqlalchemy/engine/url.py
index 1b96c3c2e1..05cc96a761 100644
--- a/lib/sqlalchemy/engine/url.py
+++ b/lib/sqlalchemy/engine/url.py
@@ -676,7 +676,10 @@ class URL(
             else:
                 name = sname
             if name is not None and getattr(self, sname, False):
-                translated[name] = getattr(self, sname)
+                value = getattr(self, sname)
+                if sname == "password":
+                    value = str(value)
+                translated[name] = value
         return translated
```

**Comment by CaselIT** [comment]
@zzzeek this does not solve the use case presented by the users, meaning that the password is dynamic. `translate_connect_args` is called only once on engine creation, breaking the above use cases

**Comment by leandropls** [comment]
Is it a viable alternative to move [`create_connect_args(u)` ](https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/engine/create.py#L564) into the [`connect`](https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/engine/create.py#L572) function?

I tend to think that, on the principle of separation of concerns, any `str()` of URL args should be done at `create_connect_args(u)`. If `create_connect_args(u)` gets called on each connect, each of the dialects could be responsible for evaluating argument strings.

It looks to me that the parsing heavy lifting is done when creating the `URL` object and that `create_connect_args` could be sufficiently lightweight.

**Comment by CaselIT** [comment]
> Is it a viable alternative to move [`create_connect_args(u)` ](https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/engine/create.py#L564) into the [`connect`](https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/engine/create.py#L572) function?

it's not very feasible, since fairly complex logic is used in some cases and/or do some imports, like
- https://github.com/sqlalchemy/sqlalchemy/blob/06f031e55f40403996e3580bdc5a22661b6303f4/lib/sqlalchemy/connectors/pyodbc.py#L45
- https://github.com/sqlalchemy/sqlalchemy/blob/06f031e55f40403996e3580bdc5a22661b6303f4/lib/sqlalchemy/dialects/mysql/mariadbconnector.py#L125 this uses an import
- https://github.com/sqlalchemy/sqlalchemy/blob/06f031e55f40403996e3580bdc5a22661b6303f4/lib/sqlalchemy/dialects/mysql/mysqldb.py#L230

Moving these in the connect path would incour a noticeable delay each time a connection is created, so I think we should avoid

**Comment by leandropls** [comment]
Loved the quagmire label 😂

Maybe we're looking at this from the wrong angle. Connecting to a cloud database service with dynamic credentials seems like a pretty reasonable use case nowadays, right? Are we using the right abstractions for this?

**Comment by CaselIT** [comment]
> Loved the quagmire label 😂

well I think this fit pretty well. Impressive how something seemingly so easy can be so complicated :)

> Maybe we're looking at this from the wrong angle. Connecting to a cloud database service with dynamic credentials seems like a pretty reasonable use case nowadays, right? Are we using the right abstractions for this?

Indeed, probably the current behaviour is not the best solution for this use case, Maybe we may want to look for possible solutions to better fit this usage?

**Comment by zzzeek** [comment]
That's the use case for str().  If you need new password on every connect , that's what this is for:. https://docs.sqlalchemy.org/en/14/core/engines.html#generating-dynamic-authentication-tokens

**Comment by francois-gay** [comment]
> That's the use case for str(). If you need new password on every connect , that's what this is for:. https://docs.sqlalchemy.org/en/14/core/engines.html#generating-dynamic-authentication-tokens

Do you know how to use this decorator with flask-sqlalchemy ?
It seems that the engine is created on the first query call, so I can't use the engine as a parameter of the decorator.
_Maybe I'm not using it the right way._

**Comment by zzzeek** [comment]
I'd report that to flask-sqlalchemy or just not use flask-sqlalchemy.   Flask can be used without it

**Comment by francois-gay** [comment]
> Do you know how to use this decorator with flask-sqlalchemy ?

Ok, it works:
```python
from flask_sqlalchemy import SQLAlchemy
from sqlalchemy import event
from flask import Falsk

app = Flask(__name__)
db = SQLAlchemy()
db.init_app(app)

engine = db.get_engine(app=app)
@event.listens_for(engine, 'do_connect')
def provide_token(dialect, conn_rec, cargs, cparams):
    cparams['token'] = get_authentication_token()
```

Thank you.

BTW: The documentation misleads

**Comment by leandropls** [comment]
> That's the use case for str(). If you need new password on every connect , that's what this is for:. https://docs.sqlalchemy.org/en/14/core/engines.html#generating-dynamic-authentication-tokens

I'm fine with doing it this way. I guess it's just a matter of removing that option (object with `__str__`) from the docs, then.

**Comment by CaselIT** [comment]
You can also register the event on the `Engine` class. If you use a single engine that's equivalent

**Comment by CaselIT** [comment]
> 
> 
> > That's the use case for str(). If you need new password on every connect , that's what this is for:. https://docs.sqlalchemy.org/en/14/core/engines.html#generating-dynamic-authentication-tokens
> 
> I'm fine with doing it this way. I guess it's just a matter of removing that option (object with `__str__`) from the docs, then.

I pointing users to that part of the documentation, adding someting like `Using an arbitrary object is not suited for dynamically changing passwords, see <link> for how to do that`

**Comment by CaselIT** [comment]
I can try implementing something for this. I guess sometimes less is more :)

**Comment by leandropls** [comment]
> Or pointing users to that part of the documentation, adding something like `Using an arbitrary object is not suited for dynamically changing passwords, see <link> for how to do that`

yeah, but this feature doesn't work anyway

**Comment by CaselIT** [comment]
> 
> 
> > Or pointing users to that part of the documentation, adding something like `Using an arbitrary object is not suited for dynamically changing passwords, see <link> for how to do that`
> 
> yeah, but this feature doesn't work anyway

well it would be fixed :)

**Comment by sqla-tester** [comment]
**Federico Caselli** has proposed a fix for this issue in the **master** branch:

**Ensure str is callect on the URL password.** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3072

**Comment by leandropls** [comment]
@CaselIT I've caught two misspells:

"When using an object with a ``__str__()`` methog" <- "methog"

"with testing.expect_deprecated_20("The names paramer is deprecated."):" <- "paramer"

I didn't really understand what "names" parameter this warning is about.

**Comment by CaselIT** [comment]
Thanks!

> I didn't really understand what "names" parameter this warning is about.

it was an old parameter that seems deprecated by more than 9 years

**Comment by leandropls** [comment]
Hey folks! I was trying to implement the event solution but it seems like there's no `do_connect` event for asyncpg:

```
  File ".../env/lib/python3.9/site-packages/sqlalchemy/event/api.py", line 142, in decorate
    listen(target, identifier, fn, *args, **kw)
  File ".../env/lib/python3.9/site-packages/sqlalchemy/event/api.py", line 99, in listen
    _event_key(target, identifier, fn).listen(*args, **kw)
  File ".../env/lib/python3.9/site-packages/sqlalchemy/event/registry.py", line 203, in listen
    dispatch_collection = getattr(target.dispatch, identifier)
  File ".../env/lib/python3.9/site-packages/sqlalchemy/event/base.py", line 106, in __getattr__
    raise AttributeError(name)
AttributeError: do_connect
```

**Comment by CaselIT** [comment]
You need to use `.sync_engine` accessort. We miss the documentation on that, as tracked by this issue https://github.com/sqlalchemy/sqlalchemy/issues/7074

**Comment by leandropls** [comment]
Got it! Thanks!
